File: ios/reactNativeNMap/RNNaverMapInfoWindow.m
//
//  RNNaverMapInfoWindow.m
//  react-native-nmap
//
//  Created by 이영준 on 2022/12/07.
//

#import "RNNaverMapInfoWindow.h"
#import "RNNaverMapMarker.h"

#define DEFAULT_INFO_WINDOW_WIDTH 100
#define DEFAULT_INFO_WINDOW_HEIGHT 44
#define DEFAULT_INFO_WINDOW_OFFSET -10

@implementation RNNaverMapInfoWindow

- (instancetype)init
{
    self = [super init];
    _realInfoWindow = [NMFInfoWindow new];
    [self createView];
    _realInfoWindow.dataSource = self;
    _zIndex = NSIntegerMin;
    _globalZIndex = NSIntegerMin;
    _borderWidth = 0;
    _offset = 0;
    
    return self;
}

- (instancetype)initWithMarker: (RNNaverMapMarker *)marker
{
    self = [self init];
    _marker = marker;
    [self setTouchHandler];
    
    return self;
}

- (void)setTouchHandler
{
    __block RNNaverMapMarker *marker = _marker;

    _realInfoWindow.touchHandler = ^BOOL(NMFOverlay *overlay) {
        if (!marker || marker.onClick == NULL)
            return false;

        marker.onClick(@{});

        return true;
    };
}

- (void)setIsVisible:(BOOL)value {
    _isVisible = value;
}

- (void)setText:(NSString *)value {
    _text = value;
}

- (void)setTextSize:(CGFloat)value {
    _textSize = value;
}

- (void)setColor:(UIColor *)value {
    _color = value;
}

- (void)setMultiline:(BOOL)value {
    _multiline = value;
}

- (void)setBackgroundColor:(UIColor *)value {
    _backgroundColor = value;
}

- (void)setBackgroundOpacity:(CGFloat)value {
    _backgroundOpacity = value;
}

- (void)setMaxWidth:(CGFloat)value {
    _maxWidth = value;
}

- (void)setPaddingHorizental:(CGFloat)value {
    _paddingHorizental = value;
}

- (void)setPaddingVertical:(CGFloat)value {
    _paddingVertical = value;
}

- (void)setCornerRadius:(CGFloat)value {
    _cornerRadius = value;
}

- (void)setZIndex:(NSInteger)value {
    _zIndex = value;
    [self applyZIndex: value];
}


- (void)setGlobalZIndex:(NSInteger)value {
    _globalZIndex = value;
    [_realInfoWindow setGlobalZIndex: value];
}

- (void)applyZIndex:(NSInteger)value {
    [_realInfoWindow setZIndex: value];
}

- (void)applyZIndexIfNeeded:(NSInteger)value {
    if (value == NSIntegerMin) return;
        
    [self applyZIndex: value];
}

- (void)setOffset:(NSInteger)value {
    _offset = value;
    [self applyOffset: value];
}

- (void)applyOffset:(NSInteger)value {
    _realInfoWindow.offsetY = DEFAULT_INFO_WINDOW_OFFSET + value;
}

- (void)applyOffsetIfNeeded {
    [self applyOffset: _offset];
}

- (void)setBorderWidth:(CGFloat)value {
    _borderWidth = value;
}

- (void)setBorderColor:(UIColor *)value {
    _borderColor = value;
}

- (void)open {
    if (_marker) {
        [_realInfoWindow openWithMarker:_marker.realMarker];
        [self applyOffsetIfNeeded];
    }
}

- (void)close {
    [_realInfoWindow close];
}

- (void)refresh {
    [_realInfoWindow invalidate];
}

- (void)createView {
    if (_view) return;
    
    _view = [[UIView alloc] initWithFrame: CGRectMake(0, 0, 100, 44)];
    _view.translatesAutoresizingMaskIntoConstraints = false;
    _view.clipsToBounds = true;
    _view.backgroundColor = [UIColor colorWithWhite: 0 alpha: 0];
    
    _backgroundView = [[UIView alloc] initWithFrame: CGRectMake(_borderWidth, _borderWidth, DEFAULT_INFO_WINDOW_WIDTH - _borderWidth * 2, DEFAULT_INFO_WINDOW_HEIGHT - _borderWidth * 2)];
    _backgroundView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _backgroundView.clipsToBounds = true;
    [_view addSubview:_backgroundView];
    
    _label = [[UILabel alloc] initWithFrame: CGRectZero];
    _label.translatesAutoresizingMaskIntoConstraints = false;
    [_view addSubview:_label];
}

- (void)updateView {
    if (!_view) return;
    
    _view.opaque = false;
    [_backgroundView setBackgroundColor: _backgroundColor];
    RCTLogTrace(@"InfoWindow updateView backgroundColor %@", _backgroundColor.description);
    [_backgroundView setAlpha: _backgroundOpacity];
    _backgroundView.layer.cornerRadius = _cornerRadius;
    
    _backgroundView.layer.borderWidth = _borderWidth;
    RCTLogTrace(@"InfoWindow updateView borderWidth %f", _borderWidth);
    
    if (_borderColor) {
        _backgroundView.layer.borderColor = _borderColor.CGColor;
    } else {
        _backgroundView.layer.borderColor = nil;
    }
    RCTLogTrace(@"InfoWindow updateView borderWidth %@", _borderColor);
    
    _label.text = _text;
    _label.font = [_label.font fontWithSize:_textSize];
    _label.textColor = _color;
    _label.numberOfLines = _multiline ? 0 : 1;
    _view.layer.cornerRadius = _cornerRadius;
    
    if (!_widthContaraint) {
        _widthContaraint = [_view.widthAnchor constraintLessThanOrEqualToConstant: _maxWidth];
        [_widthContaraint setActive: true];
    } else {
        _widthContaraint.constant = _maxWidth;
    }
    
    if (!_leftContaraint) {
        _leftContaraint = [_label.leftAnchor constraintEqualToAnchor: _view.leftAnchor constant: _paddingHorizental];
        [_leftContaraint setActive: true];
    }else {
        _leftContaraint.constant = _paddingHorizental;
    }
    
    if (!_rightContaraint) {
        _rightContaraint = [_label.rightAnchor constraintEqualToAnchor: _view.rightAnchor constant: -_paddingHorizental];
        [_rightContaraint setActive: true];
    }else {
        _rightContaraint.constant = _paddingHorizental;
    }
    
    if (!_topContaraint) {
        _topContaraint = [_label.topAnchor constraintEqualToAnchor: _view.topAnchor constant: _paddingVertical];
        [_topContaraint setActive: true];
    }else {
        _topContaraint.constant = _paddingVertical;
    }
    
    if (!_bottomContaraint) {
        _bottomContaraint = [_label.bottomAnchor constraintEqualToAnchor: _view.bottomAnchor constant: -_paddingVertical];
        [_bottomContaraint setActive: true];
    }else {
        _bottomContaraint.constant = _paddingVertical;
    }
    
    [_view layoutIfNeeded];
    [_label layoutIfNeeded];
}

- (void)destroy {
    _marker = nil;
    _realInfoWindow.touchHandler = nil;
    _realInfoWindow = nil;
}

- (nonnull UIView *)viewWithOverlay:(nonnull NMFOverlay *)overlay {
    [self updateView];
    
    return _view;
}

@end

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapInfoWindow.java
package com.github.quadflask.react.navermap;

import android.content.Context;
import android.graphics.drawable.GradientDrawable;
import android.text.TextUtils;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.naver.maps.map.overlay.InfoWindow;

public class RNNaverMapInfoWindow {
    /**
     * 원래 0이 기본 값이지만 0일 때 간격이 없는 상태가 아님
     */
    static final int DEFAULT_INFO_WINDOW_OFFSET = -11;

    boolean isVisible = false;
    String text;
    Double textSize;
    Integer color;
    boolean multiline = false;
    Integer backgroundColor;
    Double backgroundOpacity = 1d;
    Integer maxWidth;
    Integer paddingHorizental = 0;
    Integer paddingVertical = 0;
    Integer cornerRadius = 8;
    Integer zIndex;
    Integer globalZIndex;
    Integer offset;
    Integer borderColor;
    Integer borderWidth;

    InfoWindow.ViewAdapter adapter;
    TextView textView;
    RelativeLayout textViewContainer;

    private InfoWindow infoWindow;
    private RNNaverMapMarker marker;

    float pixelRatio = 1;

    public RNNaverMapInfoWindow() {
        this.infoWindow = new InfoWindow();
    }

    public RNNaverMapInfoWindow(RNNaverMapMarker marker) {
        this.marker = marker;
        this.infoWindow = new InfoWindow();
        this.infoWindow.setOnClickListener(marker);

        this.createAdapter();
    }

    public RNNaverMapInfoWindow setIsVisible(boolean value) {
        this.isVisible = value;
        return this;
    }

    public RNNaverMapInfoWindow setText(String value) {
        if (value == null) return this;

        this.text = value;
        return this;
    }

    public RNNaverMapInfoWindow setTextSize(Double value) {
        if (value == null) return this;

        this.textSize = value;
        return this;
    }

    public RNNaverMapInfoWindow setColor(Integer value) {
        if (value == null) return this;

        this.color = value;
        return this;
    }

    public RNNaverMapInfoWindow setMultiline(boolean value) {
        this.multiline = value;
        return this;
    }

    public RNNaverMapInfoWindow setBackgroundColor(Integer value) {
        if (value == null) return this;

        this.backgroundColor = value;
        return this;
    }

    public RNNaverMapInfoWindow setBackgroundOpacity(Double value) {
        if (value == null) return this;

        this.backgroundOpacity = value;
        return this;
    }

    public RNNaverMapInfoWindow setMaxWidth(Integer value) {
        if (value == null) return this;

        this.maxWidth = value;
        return this;
    }

    public RNNaverMapInfoWindow setPaddingHorizental(Integer value) {
        if (value == null) return this;

        this.paddingHorizental = value;
        return this;
    }

    public RNNaverMapInfoWindow setPaddingVertical(Integer value) {
        if (value == null) return this;

        this.paddingVertical = value;
        return this;
    }

    public RNNaverMapInfoWindow setCornerRadius(Integer value) {
        if (value == null) return this;

        this.cornerRadius = value;
        return this;
    }

    public RNNaverMapInfoWindow setZIndex(Integer value) {
        if (value == null) return this;

        this.zIndex = value;
        this.applyZIndex(value);
        return this;
    }

    public RNNaverMapInfoWindow setGlobalZIndex(Integer value) {
        if (value == null) return this;

        this.globalZIndex = value;
        this.infoWindow.setGlobalZIndex(value);
        return this;
    }

    public void applyZIndex(int value) {
        this.infoWindow.setZIndex(value);
    }

    public RNNaverMapInfoWindow setOffset(Integer value) {
        this.offset = value;
        this.applyOffset(value);

        return this;
    }

    public void applyOffset(Integer value) {
        Integer offsetWithDefault = value != null ? DEFAULT_INFO_WINDOW_OFFSET + value : null;
        int pixelsInOffset = offsetWithDefault != null ? ((Double) (offsetWithDefault.doubleValue() * pixelRatio)).intValue() : 0;

        this.infoWindow.setOffsetY(pixelsInOffset);
    }

    public RNNaverMapInfoWindow setBorderColor(Integer value) {
        if (value == null) return this;

        this.borderColor = value;
        return this;
    }

    public RNNaverMapInfoWindow setBorderWidth(Integer value) {
        if (value == null) return this;

        this.borderWidth = value;
        return this;
    }

    public void open() {
        if (this.marker == null) return;

        this.infoWindow.open(this.marker.feature);
    }

    public void refresh() {
        this.infoWindow.invalidate();
    }

    public  void close() {
        this.infoWindow.close();
    }

    public void createAdapter() {
        if (this.marker == null) return;

        this.adapter = new InfoWindow.ViewAdapter() {
            @NonNull
            @Override
            public View getView(@NonNull InfoWindow infoWindow) {
                if (textViewContainer == null) {
                    textViewContainer = new RelativeLayout(marker.getContext());
                }

                if (textView == null) {
                    textView = new TextView(marker.getContext());
                    textViewContainer.addView(textView);
                }

                pixelRatio = marker.getContext().getResources().getDisplayMetrics().density;

                if (text != null) textView.setText(text);
                if (textSize != null) textView.setTextSize(textSize.floatValue());
                if (color != null) textView.setTextColor(color);
                textView.setSingleLine(!multiline);
                textView.setEllipsize(TextUtils.TruncateAt.END);
                if (maxWidth != null) {
                    Double maxWidthWithPadding = maxWidth.doubleValue() + paddingHorizental * 2;
                    textView.setMaxWidth(((Double)(maxWidthWithPadding * pixelRatio)).intValue());
                }
                int pixelsInPaddingHorizental = ((Double)(paddingHorizental.doubleValue() * pixelRatio)).intValue();
                int pixelsInPaddingVertical = ((Double)(paddingVertical.doubleValue() * pixelRatio)).intValue();
                textViewContainer.setPadding(pixelsInPaddingHorizental, pixelsInPaddingVertical, pixelsInPaddingHorizental, pixelsInPaddingVertical);

                int pixelsInRadius = ((Double)(cornerRadius.doubleValue() * pixelRatio)).intValue();

                if(backgroundColor != null) {
                    GradientDrawable gd = new GradientDrawable();
                    gd.setColor(backgroundColor);
                    gd.setCornerRadius(pixelsInRadius);

                    if (borderWidth != null && borderColor != null) {
                        gd.setStroke(borderWidth, borderColor);
                    }
                    gd.setAlpha((int)(backgroundOpacity * 255));
                    textViewContainer.setBackground(gd);
                }

                applyOffset(offset);

                return  textViewContainer;
            }
        };

        this.infoWindow.setAdapter(this.adapter);
    }

    public void destroy() {
        this.marker = null;
        if (this.infoWindow != null) return;
        this.infoWindow.setOnClickListener(null);
        this.infoWindow.setAdapter(null);
        this.infoWindow = null;
    }
}
